lib/Sema/CSSimplify.cpp

```diff
@@ -430,24 +430,24 @@ matchCallArguments(SmallVectorImpl<AnyFunctionType::Param> &args,
 
   // If we have a trailing closure, it maps to the last parameter.
   if (hasTrailingClosure && numParams > 0) {
-    unsigned lastParamIdx = numParams - 1;
-    bool lastAcceptsTrailingClosure =
-        acceptsTrailingClosure(params[lastParamIdx]);
-
     // If the last parameter is defaulted, this might be
     // an attempt to use a trailing closure with previous
     // parameter that accepts a function type e.g.
     //
     // func foo(_: () -> Int, _ x: Int = 0) {}
     // foo { 42 }
-    if (!lastAcceptsTrailingClosure && numParams > 1 &&
-        paramInfo.hasDefaultArgument(lastParamIdx)) {
-      auto paramType = params[lastParamIdx - 1].getPlainType();
-      // If the parameter before defaulted last accepts.
-      if (paramType->is<AnyFunctionType>()) {
+    bool lastAcceptsTrailingClosure = false;
+    unsigned lastParamIdx = numParams - 1;
+    for (unsigned idx = numParams - 1;; --idx) {
```

Comment:
Hm that doesn’t really make it cleaner either... How about iterating over params directly and decreasing an index instead?

Reply:
Do you mean that eliminate `idx` and decrement `lastParamIdx` directly?
If there is no function parameter, `lastParamIdx` needs to keep `numParams - 1`.
It introduce additional `if` statement after loop.
So code will be:

```
unsigned lastParamIdx = numParams - 1;
for (;; --lastParamIdx) {
  if (acceptsTrailingClosure(params[lastParamIdx])) {
        lastAcceptsTrailingClosure = true;
        break;
  }
  if (!paramInfo.hasDefaultArgument(lastParamIdx))
        break;
  if (!lastParamIdx)
        break;
}
if (!lastAcceptsTrailingClosure)
  lastParamIdx = numParams - 1;
```

Reply:
Yeah, that doesn't make it better either :/ Let's go with the original version where it used to iterate forward over parameters, to make it a little more concise though you could use `for (auto i : indices(params)) {`